### DISCUSSION TITLE
SDK API design question: why "client.request" requires a **function** as unique argument ?

### DISCUSSION THREAD

**Discussion by pomeh** [directus/directus#23147] [Feature Requests]
### Summary

I don't understand why the `client.request` requires a **function** as argument, as this complexifies some code when you want to customize the request for example.

### Basic Example

Using the JS SDK and according to the documentation, I need to use this syntax to get items from a collection:

```ts
    client.request(readItems('foo', {
        fields: [ 'id', 'slug', /* ... */ ],
    }));
```

For some reasons, I need to edit the request before sending it, so I need to extend the result from `readItems` before passing it to `client.request` func.

Basically, I would have done is this way:

```ts
    // 1. build the initial request
    const req = readItems('foo', {
        fields: ['id', 'slug', /* ... */],
    });
 
    // 2. extend the request
    req.headers = { "X-Custom": "yes" };

    // 3. send the request
    client.request(req);
```

It turns out, the code above **does not work**, and it seems it's because of a simple reason: `request` func accepts a **function** as argument, and not an **object**.

Out of curiosity, I look out the [`request` source code](https://github.com/directus/directus/blob/c4a2f18381205579df7f33c7381e29ee2c88be81/sdk/src/rest/composable.ts#L19C5-L19C34) to understand why it was designed this way:

```ts
			async request<Output = any>(getOptions: RestCommand<Output, Schema>): Promise<Output> {
				const options = getOptions();
```

But as we can see, we only call the `getOptions` function as is, without any added argument or anything else, so I guess having an object in the first place would have ended with the same result.

### Motivation

By accepting an object instead of a function would allow for a code more simple in my use case.

With such a design, this code would work:

```ts
// 1. build the initial request
const req = readItems('foo', {
    fields: ['id', 'slug', /* ... */],
});

// 2. extend the request
req.headers = { "X-Custom": "yes" };

// 3. send the request
client.request(req);
```

But currently, I have to write more code to achieve the same goal:

```ts
// limitation one: you have to explicitely use the <Schema, ...> syntax for the `readItems` arguments types to be correct
const builder = readItems<Schema, 'foo', Query<Schema, CollectionType<Schema, 'foo'>>>('foo', {
    fields: ['id', 'slug', /* ... */],
}); 

// limitation two: you have to call the `readItems` result by yourself
const req = builder();

// extend the request
req.headers = { "X-Custom": "yes" };

// limitation three: you have to define a new anonymous function to return the now extended plain JS object
return client.request<Schema>(() => req);
```

### Detailed Design

1. Ability to extend `readItems` value:
 
```ts
client.request<Schema>(() => ({
    ...readItems('foo', {
         fields: ['id', 'slug', /* ... */],
    }),
    headers: {
        "X-Custom": "yes",
    }
}));
```

2. Ability to use plain old JavaScript object of your own:

```ts
client.request<Schema>({
    path: '/items/foo',
    method: 'GET',
    headers: {
        "X-Custom": "yes",
    }
});
```

### Must Have:
- `client.request` method must accept a plain old JavaScript object as argument

### Could Have:
- `client.request` method could accept both function and plain old JavaScript object as argument, for better code migration path

### Drawbacks

Implementing this (switch from function to object arg), would require to change `readItems` method (& similar) return type and removing the line `const options = getOptions();` from the `request` function.

As far as I know, this look like a small change to do in the Directus SDK code, but I know very little about it and I'm sure to miss something :)

The huge drawback here is this would be a breaking change in the SDK API design, thus require all Directus dev-users to adapt their code for this.

This last drawback could be minimized by allowing both function and object as argument.

### Alternatives

The only alternative I can think of is doing it how I explained above, but with the same impacts on the code quality and maintenance cost:

```ts
const builder = readItems<Schema, 'foo', Query<Schema, CollectionType<Schema, 'foo'>>>('foo', {
    fields: ['id', 'slug', /* ... */],
}); 

const req = builder();

req.headers = { "X-Custom": "yes" };

return client.request<Schema>(() => req);
```

### Adoption Strategy

for `request` method:

1. support both function and object, but function syntax beeing deprecated
2. remove function syntax in a later version. This would be a breaking change

for `readItems` (& similar) method:

I'm not sure about this.

Stategy one:

1. while `request` accept both syntax, do not change `readItems` syntax. This will trigger deprecation notice to developpers, and there's not much they could do to fix it.
2. switch to object return type when `request` remove function syntax. This would be a breaking change too

Stategy two:

1. switch to object return type as soon as `request` support both syntax, but this would be an additional breaking change

**Comment by br41nslug** [comment]
The `client.request` takes a function as parameter not so much for functional reasons but mainly to support generic typescript types. Using generics the individual endpoints (`readItems` etc) can infer the provided `Schema` from their parent to provide type hinting/completion on the endpoints inputs as well actually do run-time checks against this input if needed (check for empty arrays for example) and last but not least a way to define the varying output types for different endpoints.

> 2. Ability to use plain old JavaScript object of your own:

You can do this by wrapping the object in an arrow function like so:
```ts
client.request(() => ({
    path: '/items/foo',
    method: 'GET',
    headers: {
        "X-Custom": "yes",
    }
}));
```
this allows users heavily relying on typescript types to author their own functions with inferred types where a plain object does not have this option.

> 1. Ability to extend readItems value:

We provide typed helper functions to extend endpoint definitions like:
```ts
client.request(withOptions(
    readItems("foo", {
      fields: ["id", "slug" /* ... */],
    }),
    {
      headers: {
        "X-Custom": "yes",
      },
    }
));
```

I would love to see some examples/suggestions on how this object format could support input and output typing. For example:
```ts
// how would "result" be typed here based on the object (supporting all the various endpoints in the api)
const result = await client.request({
    path: '/items/foo', // how about type hinting a valid api path (like we do with `readItems` collection parameter)
    method: 'GET',
    headers: {
        "X-Custom": "yes",
    }
}));
```
